addons/sale_timesheet_margin/models/sale_order_line.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import api, models
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.depends('analytic_line_ids.amount')
+    def _compute_purchase_price(self):
+        timesheet_sol = self.filtered(lambda sol: sol.qty_delivered_method == 'timesheet')
+        super(SaleOrderLine, self - timesheet_sol)._compute_purchase_price()
+        if timesheet_sol:
```

Comment:
In the case where we have some records in `timesheet_sol` and all of those records have no product_id, we do a unnecessary `read_group` 😢. To avoid this we could add a condition in your `if` statement.
```suggestion
        if timesheet_sol and any(line.product_id for line in timesheet_sol):
```

Reply:
Is it possible to have a sol.qty_delivered_method without product ?